answer2_Swift.playground/Contents.swift

```diff
@@ -0,0 +1,253 @@
+ 
+ Step0: Complete LRU Cache
+ */
+//:----
+//: ## LRU Cache
+//: ### Analyze
+/*
+ LRU Cache retain a sorted and finite amount of data. The data is sorted in the order by the time it been accessed.
+ When the capacity is full, the least recently used object of data will be delete.
+ 
+ Step0: A payload protocol to contain key and value for node.
+ Step1: CachePayload for node to store.
+ Step2: Node contain payload, previous link and next link.
+ Step3: Doubly Linked List insert, remove, MoveTo
+ */
+//:----
+protocol Payload {
+    associatedtype Key
+    associatedtype Value
+    
+    var key: Key { get set }
+    var value: Value { get set }
+}
+
+struct CachePayload<T: Hashable, U>: Payload {
+    
+    var key: T
+    var value: U
+    
+    init(key: Key, value: Value) {
+        self.key = key
+        self.value = value
+    }
+}
+
+class Node<T: Payload> {
+    
+    var payload: T
+    var previous: Node<T>?
+    var next: Node<T>?
+    
+    init(value: T) {
+        self.payload = value
+    }
+}
+
+class DoublyLinkedList<T: Payload> {
+    
+    // The head starting at index 0.
+    var head: Node<T>?
+    var tail: Node<T>?
+    var isEmpty: Bool {
+        return head == nil && tail == nil
+    }
+    
+    public var count: Int = 0
+    
+    public func insertHead(node: Node<T>) {
+        guard !isEmpty else {
+            head = node
+            tail = node
+            count += 1
+            return
+        }
+        
+        var currNode = head
+        
+        if currNode === head {
+            if head === tail {
+                node.next = head
+                head?.previous = node
+                head = node
+            } else {
+                node.next = head
+                head = node
+            }
+            
+            count += 1
+            return
+        }
+        
+        node.previous = currNode
+        node.next = currNode?.next
+        currNode?.next?.previous = node
+        currNode?.next = node
+        
+        count += 1
+        return
+    }
+    
+    func removeTail() {
+        guard !isEmpty else {
+            return
+        }
+        
+        if head === tail {
+            head = nil
+            tail = nil
+            count -= 1
+            return
+        }
+        
+        tail?.previous?.next = nil
+        tail = tail?.previous
+        
+        count -= 1
+        return
+    }
+    
+    public func moveToHead(node: Node<T>) {
+            guard !isEmpty else {
+                return
+            }
+
+            if head === node && tail === node {
+                // do nothing
+            } else if head === node {
+                // do nothing
+            } else if tail === node {
+                tail?.previous?.next = nil
+                tail = tail?.previous
+
+                let prevHead = head
+                head?.next?.previous = node
+                head = node
+                head?.next = prevHead
```

Comment:
可以複用前面的方法（假設前面方法有寫對）

```suggestion
                let tail = removeTail()
                insertHead(node: tail)
```